docs/plutus-report/scripting.tex

```diff
@@ -111,7 +115,127 @@ \subsection{Recursive values}
+  But we don't want this functionality on-chain, so we want to remove the builtin (or replace it with a no-op).
+\item
+  We may need different builtins for different ledgers (see \cref{req:script-lang-extensibility}).
+\end{enumerate}
+
+So we need to make builtin types and functions extensible.
+
+\subsubsection{Builtin types}
+
+Extensible builtin types are relatively straightforward on the type level, but become trickier on the value level.
+Values of builtin types are \emph{constants}, but what is a constant?
+It's easy to say what a \code{Integer} constant should be: an integer!
+However, what about constants of extensible builtin types?
+
+We adopt an approach familiar from generic programming in dependently-typed languages, where we parameterize our language over a \emph{universe} of types.\footnote{
+  Where do these types themselves come from?
+  We assume that we are working in some ``metalangauge'' which has types such as \code{Integer} that we can reuse in Plutus Core.
+}
+A constant then is then \emph{tagged} with a value corresponding to a type in that universe, and a value of that type.
+
+This does complicate interpretation of \gls{plutus-core} programs, since in order to interpret a program, you also need to know what universe of types it is expecting.
+In practice this is not so bad: for most users (e.g. the \gls{cardano} ledger) there will be a single universe that they use, and from there on they can simply act as though we did not have extensible builtins.
+
+\paragraph{Builtin datatypes}
+
+Adding new types to \gls{plutus-core} is not very useful unless they have functions that can operate on them!
+A particularly common case is where the new type is (morally) a datatype, and should support pattern matching.
+We can support this by also providing a ``matching'' function as a builtin function.\footnote{
+  These functions look just like the ``matchers'' for Scott-encoded datatypes, which is no coincidence.
+}
+
+For example, the matching function for \code{bool} is \code{ifThenElse :: forall a . bool -> a -> a -> a}.
+Matching functions are always polymorphic, which is a key reason why we need to support polymorphic builtin functions (see \cref{sec:polymorphic-builtins}).
+
+\paragraph{Polymorphic builtin types}
+
+It would be nice to support polymorphic builtin types.
+For example, that would allow us to provide optimized implementations of container types such as maps.
+
+This seems feasible in principle, but we have not worked the details out yet.
+
+\subsubsection{Builtin functions}
+
+We could implement extensible builtin functions in a similar way to extensible builtin types: use some universe that catalogues the available builtin functions.
+However, this is increasingly tedious, and we don't have the same dependency on the available functions as we do on the types, since we don't have anything corresponding to constants.
```

Comment:
What is increasingly tedious? In the end we do use a very similar GADT approach. I guess, I just don't get what this entire subsection is saying, i.e. what exactly the other approach that we could go with is.

Reply:
I was gesturing at the issues you raise with approach 2 here: https://github.com/input-output-hk/plutus/pull/2001/files#diff-d31de035f5939b8a54e0c49ba59b2f8fR160

I'm actually not sure I agree that we should eventually go with the extra type parameter that constrains the builtin functions at compile time, essentially I'm arguing that we should stick with approach 1.

Reply:
> I was gesturing at the issues you raise with approach 2 here

I see.

> I'm actually not sure I agree that we should eventually go with the extra type parameter that constrains the builtin functions at compile time, essentially I'm arguing that we should stick with approach 1.

Ah, I didn't realize that was the long-term plan. I'll fix the builtins writeup.

Reply:
> Ah, I didn't realize that was the long-term plan.

I don't think we know! It's my best guess at the moment, but if we decide to do it the other way I'll just update this :)